pinot-core/src/main/java/org/apache/pinot/core/util/ListenerConfigUtil.java

```diff
@@ -0,0 +1,206 @@
+        getPort(config.getProperty(protocolNamespace + ".port")), protocol, tlsConfig);
+  }
+
+  private static String getHost(String configuredHost) {
+    return Optional.ofNullable(configuredHost).filter(host -> !host.trim().isEmpty())
+        .orElseThrow(() -> new IllegalArgumentException(configuredHost + " is not a valid host"));
+  }
+
+  private static int getPort(String configuredPort) {
+    return Optional.ofNullable(configuredPort).filter(port -> !port.trim().isEmpty()).<Integer> map(Integer::valueOf)
+        .orElseThrow(() -> new IllegalArgumentException(configuredPort + " is not a valid port"));
+  }
+
+  public static HttpServer buildHttpServer(ResourceConfig resConfig, List<ListenerConfig> listenerConfigs) {
+    Preconditions.checkNotNull(listenerConfigs);
+
+    // The URI is irrelevant since the default listener will be manually rewritten.
+    HttpServer httpServer = GrizzlyHttpServerFactory.createHttpServer(URI.create("http://0.0.0.0/"), resConfig, false);
+
+    // Listeners cannot be configured with the factory. Manual overrides are required as instructed by Javadoc.
+    httpServer.removeListener("grizzly");
+
+    listenerConfigs.forEach(listenerConfig -> configureListener(httpServer, listenerConfig));
+
+    try {
+      httpServer.start();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to start http server", e);
+    }
+
+    return httpServer;
+  }
+
+  public static void configureListener(HttpServer httpServer, ListenerConfig listenerConfig) {
+    final NetworkListener listener = new NetworkListener(listenerConfig.getName() + "-" + listenerConfig.getPort(),
+        listenerConfig.getHost(), listenerConfig.getPort());
+
+    listener.getTransport().getWorkerThreadPoolConfig()
+        .setThreadFactory(new ThreadFactoryBuilder().setNameFormat("grizzly-http-server-%d")
+            .setUncaughtExceptionHandler(new JerseyProcessingUncaughtExceptionHandler()).build());
+
+    if (listenerConfig.getTlsConfig().isEnabled()) {
+      listener.setSecure(true);
+      listener.setSSLEngineConfig(buildSSLConfig(listenerConfig.getTlsConfig()));
+    }
+
+    httpServer.addListener(listener);
+  }
+
+  private static SSLEngineConfigurator buildSSLConfig(TlsConfig tlsConfig) {
+    SSLContextConfigurator sslContextConfigurator = new SSLContextConfigurator();
+
+    sslContextConfigurator.setKeyStoreFile(tlsConfig.getKeyStorePath());
+    sslContextConfigurator.setKeyStorePass(tlsConfig.getKeyStorePassword());
+    sslContextConfigurator.setTrustStoreFile(tlsConfig.getTrustStorePath());
+    sslContextConfigurator.setTrustStorePass(tlsConfig.getTrustStorePassword());
+
+    return new SSLEngineConfigurator(sslContextConfigurator).setClientMode(false)
+        .setNeedClientAuth(tlsConfig.isClientAuthEnabled()).setEnabledProtocols(new String[] { "TLSv1.2" });
+  }
+
+  public static String toString(Collection<? extends ListenerConfig> listenerConfigs) {
```

Comment:
Why not move this to ListenerConfig class?

Reply:
It takes a collection. imo keep the POJOs as clean as possible.